rfd/0048-desktop-access-session-recording.md

```diff
@@ -0,0 +1,300 @@
+- message 6 - clipboard data: to indicate when the clipboard was used
+- message 3 - mouse move
+- message 4 - mouse button
+
+Of note: we will not be capturing other user input events (keyboard or mouse
+wheel/scroll), as they are not necessary for playback. The result of keyboard
+input or scrolling captured in the session's PNG frames instead.
+
+We do, however, need to capture mouse movement and button clicks, as the RDP
+bitmaps sent from the Windows desktop do _not_ include the mouse cursor
+(RDP clients show the native OS cursor and send its position to the desktop).
+
+When this new event is introduced, several parts of the code that assume
+`SessionPrintEvent` is the only type of session recording will need to be
+udpated:
+
+- `AuditWriter` assumes every `[]byte` it receives needs to be packaged up into
+  a `SessionPrintEvent`. We will extend `AuditWriterConfig` with a new field:
+  `MakeEvent func(b []byte) AuditEvent`. This will enable SSH sessions to
+  configure `AuditWriter` to produce `SessionPrintEvent`s, while allowing
+  desktop sessions to emit `DesktopRecordingEvent`s.
+- `AuditWriter` does some tracking of `lastPrintEvent` in order to update
+  timestamps and chunk indicdes. This needs to be generalized to be able to
+  track the last event in desktop mode as well.
+- The `TeeStreamer` that filters out `SessionPrintEvent`s needs to be updated to
+  also filter out `DesktopRecordingEvent`s, as there will be a large number of
+  these events and we don't want to clutter the global audit log with them.
+
+##### Recording Mode
+
+SSH session recording can operate in synchronous or asynchronous mode, and can
+be captured at the node or at the proxy, resulting in a total of 4 possible
+configurations.
+
+- `node`
+- `node-sync`
+- `proxy`
+- `proxy-sync`
+- `off`
+
+Synchronous recording mode emits each audit event directly to the audit log, and
+fails if an event can not be written. Asynchronous recording mode writes events
+to a local filesystem log, and periodically uploads the audit events to external
+storage. Asynchronous mode is both more efficient (due to making less API calls)
+and more resilient (due to support for retries, backoff, stream recovery, etc).
+For desktop recordings, Teleport will support sync or async modes.
+
+Unlike SSH sessions, where the node itself is running Teleport and the user has
+an option of where the recording will take place, Desktop Access sessions will
+always be recorded by the Windows Desktop Service which has an RDP connection to
+the Windows Host. This is because we only need to record certain types of
+messages, and the Windows Desktop Service is the only part of the stack that is
+TDP protocol-aware (the Teleport Proxy service simply passes data between the
+browser and the Windows Desktop Service without attempting to decode or
+interpret the data in any way.)
+
+To summarize, desktop session recording will interpret the session recording
+configuration as follows:
```

Comment:
@xinding33 interested in your feedback here. I'm trying to tap into the existing configuration as much as possible to avoid having separate recording configurations for desktop and SSH.